🧰  💾 Add utilities to analyse datasets

mberr opened the pull request:
This PR adds some utilities to analyse datasets.

Currently they mainly revolve around entity/relation frequencies.

mberr committed:
Add factory_dict property to DataSet

mberr committed:
Add first draft of analysis utilities

mberr committed:
Improve dataset description

mberr committed:
Add utility to create entity-relation co-occurrence dataframe

mberr committed:
Merge remote-tracking branch 'public-pykeen/master' into add-dataset-analysis-utilities

mberr committed:
Add newline to end of file utils.py

mberr committed:
Fix flake8:B007

mberr committed:
Fix missing module docstring

mberr committed:
Fix blank line with whitespace

mberr committed:
Fix "line too long"

cthoyt commented on src/pykeen/datasets/analysis.py:
```diff
@@ -0,0 +1,217 @@
+import pandas
+import torch
+
+from pykeen.datasets import DataSet
```
relative imports, please :)

cthoyt commented on src/pykeen/datasets/analysis.py:
```diff
@@ -0,0 +1,217 @@
+) -> Mapping[str, Any]:
+    """Describe a tensor of IDs.
+
+    In particular computes
```
missing end of sentence?

cthoyt committed:
Tiny style changes

- don't need parentheses inside indexes
- if it all fits on one line, don't break function definitions
- itemgetterrrrr

[skip ci]

mberr committed:
Use relative imports

mberr committed:
Add invert_mapping to __all__

mberr committed:
Improve docstring

mberr committed:
Fix typing warning

mberr committed:
Improve docstring

mberr committed:
Use view instead of torch.cat

mberr committed:
Improve documentation

mberr committed:
Fix typo

mberr committed:
Improve documentation

mberr committed:
Merge remote-tracking branch 'origin/master' into add-dataset-analysis-utilities

# Conflicts:
#	src/pykeen/utils.py

mberr committed:
Fix import

mberr committed:
Add relation count dataframe

mberr committed:
Consistent dataframe format

mberr committed:
Add unittest

mberr committed:
Remove non-dataframe analysis code

mberr committed:
Fix canonical order

mberr committed:
Add example usage to docstring

mberr committed:
Improve docstring

mberr committed:
Add example usage to entity_count_dataframe

mberr committed:
Add example to docstring

mberr committed:
Fix line-too-long

[skip ci]

mberr committed:
Fix line-too-long

[skip ci]

mberr committed:
Fix docstring format

cthoyt commented on src/pykeen/utils.py:
```diff
@@ -440,3 +441,20 @@ def ensure_ftp_directory(*, ftp: ftplib.FTP, directory: str) -> None:
+    num_keys = len(mapping)
+    if num_unique_values < num_keys:
+        raise ValueError(f'Mapping is not bijective! Only {num_unique_values}/{num_keys} are unique.')
+    return dict(map(reversed, mapping.items()))
```
maybe `{v:k for k,v in mapping.items()}` would get the point across a bit easier

lvermue committed:
Merge branch 'master' into add-dataset-analysis-utilities

cthoyt committed:
Merge branch 'master' into add-dataset-analysis-utilities

cthoyt committed:
Update base.py

mberr committed:
Add first draft of relation classification

mberr committed:
Merge remote-tracking branch 'origin/add-dataset-analysis-utilities' into add-dataset-analysis-utilities

mberr committed:
fix import error

mberr committed:
allow restriction of patterns

mberr committed:
Add test on bigger dataset

mberr committed:
move pandas-isation into method

mberr committed:
re-organize code

mberr committed:
fix anti-symmetry?

mberr committed:
add comment about anti-symmetry

mberr committed:
some first steps towards caching

mberr committed:
Remove unused method

mberr committed:
re-order method

mberr committed:
Merge pattern calculation again

mberr committed:
Make sure cache directory exists

mberr committed:
Update test

mberr committed:
update comment

mberr committed:
Add more comments

mberr committed:
Add type annotation for temporary index structures

mberr committed:
Extract utility method

mberr committed:
update docstring

mberr committed:
re-order code

mberr committed:
split into methods

mberr committed:
update imports

mberr committed:
flake8 fixes

Trigger CI

mberr committed:
Improve some docstrings

cthoyt commented on src/pykeen/datasets/analysis.py:
```diff
@@ -0,0 +1,449 @@
+    :param data_stream:
+        The stream of data, comprising tuples (relation_id, pattern-type, support, confidence).
+
+    :return:
```
@mberr you want to write instead 
```
yields: entries in the support-confidence skyline.
```
and then describe a single object that gets yielded (since it's implied several will be yielded by the type annotation)

mberr replied:
with or without another `:` in front? :sweat_smile:

cthoyt replied:
my b. That was a typo. Use the `:yields:`

cthoyt committed:
Cleaner itertools usage

cthoyt committed:
Update tqdm import

cthoyt committed:
Make logger come first

mberr committed:
update docstring for yield

cthoyt commented on src/pykeen/datasets/analysis.py:
```diff
@@ -301,8 +301,8 @@ def _yield_binary_patterns(
-    :return:
-        An iterable of (relation_id, pattern_type, support, confidence) tuples.
+    :yields:
+        A tuple (relation_id, pattern_type, support, confidence).
```
if you want really fine granular documentation for these kinds of iterables of tuples, you can always create a named tuple class as a subclass of `typing.NamedTuple` and document the fields there ;)

cthoyt commented on src/pykeen/datasets/analysis.py:
```diff
@@ -0,0 +1,449 @@
+    Yield ternary patterns from pre-indexed triples.
+
+    Patterns:
+        Composition
```
can we also do transitivity? 

`r(x, y) \land r(y, z) \implies r(x, z)`

cthoyt replied:
or also anti-transitivity (I made this word up)

`r(x, y) \land r(y, z) \implies r'(x, z)`

Example: X decreases Y. Y decreases Z. Imply X increases Z.

mberr replied:
I guess so :sweat_smile: For now, I added only the ones from the RotatE paper, also requested by one of the reviewers of the benchmarking paper.

However, the "data structure" (relation_id, pattern, support, confidence) allows relatively easy extension to other patterns.

mberr committed:
Merge branch 'master' into add-dataset-analysis-utilities

cthoyt committed:
Update docs and type annotations

cthoyt committed:
Add named tuple and update type annotations

cthoyt committed:
Take advantage of named tuple dot access

cthoyt committed:
Add TODO

mberr committed:
Move logging into methods

mberr committed:
Allow to select parts

mberr committed:
fix hashing

mberr committed:
fix hashing again

mberr committed:
Add force option

mberr committed:
Improve tqdm description

Trigger CI

cthoyt commented on src/pykeen/datasets/analysis.py:
```diff
@@ -0,0 +1,484 @@
+    parts = [parts] if isinstance(parts, str) else parts
+
+    # include part hash into cache-file name
+    part_hash = hashlib.sha512("".join(sorted(parts)))[:16]
```
alternate concept for part_hash:

```python
str(int('training' in parts)) + str(int('testing' in parts)) + str(int('validation' in parts))
```

this would just be a string of 0/1 three times in a row

mberr replied:
The hash is not good enough yet, since it does not cover different split sizes :confused:

cthoyt replied:
ooooh interesting, maybe we need a generalize hash for a given dataset that represents which triples ended up in testing, training, and validation?

mberr replied:
For now I went with

```python
"".join(map(str, sorted(mapped_triples.tolist()))).encode("utf8")
```

cthoyt committed:
Merge branch 'master' into add-dataset-analysis-utilities

mberr committed:
compute hash over sorted triple set

Trigger CI

mberr committed:
Merge remote-tracking branch 'origin/add-dataset-analysis-utilities' into add-dataset-analysis-utilities

mberr committed:
Extract utility method

PyKEEN-bot committed:
Trigger CI

mberr committed:
fix flake8

Trigger CI

mberr committed:
Merge remote-tracking branch 'origin/add-dataset-analysis-utilities' into add-dataset-analysis-utilities

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Update factory_dict

cthoyt committed:
Code cleanup

The hash protocol doesn't work on old versions of python < 39

Trigger CI

cthoyt committed:
Update formatting

cthoyt committed:
Update iteration

cthoyt committed:
Improve performance of skyline

If you pre-sort then you can skip a bunch of checks. Still need to do some grouping so we can avoid the first s2>s check

mberr commented on src/pykeen/datasets/analysis.py:
```diff
@@ -191,13 +191,13 @@ def entity_relation_co_occurrence_dataframe(dataset: Dataset) -> pd.DataFrame:
-            s2 > s and c2 > c
-            for s2, c2 in xs
+            s2 > s and c2 >= c
+            for s2, c2 in xs[i:]
```
Still `O(n^2)` isn't it? :smile: But besides, might be an improvement in runtime.

cthoyt replied:
yes, true. but this will half the runtime ;)

I have been staring at this for like 80 minutes trying to come up with something better. I know there are algorithms for this. They're totally non-intuitive but super elegant. It's a reminder I can't come up with clever algorithms for 💩  these days

cthoyt replied:
https://stackoverflow.com/questions/19059878/dominant-set-of-points-in-on < sort of relevant, but this finds the dominant points, not the ones that aren't dominated (unless that's probably the same)

mberr replied:
I think this one is applicable, cf. db2c4fd

mberr committed:
update skyline computation

mberr committed:
Merge branch 'master' into add-dataset-analysis-utilities

PyKEEN-bot committed:
Trigger CI

mberr committed:
adjust old skyline

Trigger CI

cthoyt committed:
Merge branch 'master' into add-dataset-analysis-utilities

cthoyt committed:
Fix yields docs

@mberr you can't do a newline here before the explanation

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Update analysis.py

Trigger CI

mberr committed:
Remove test for FB15k dataset due to excessive runtime

PyKEEN-bot committed:
Trigger CI
